### PR TITLE
Support installation in a subdirectory

### LINKED ISSUE #687: Support installation in a subdirectory
PufferPanel 0.7 could be configured to be accessed from a subdirectory (e.g. https://example.com/pufferpanel/), however PufferPanel 0.8 fails for a variety of reasons in this situation, even when "PufferPanel Master URL" is set correctly.

From what I can see, in order to resolve the issue,
1. The notes on [this page](https://github.com/klein/klein.php/wiki/Sub-Directory-Installation) must be implemented.
2. Calls to `$response->redirect` must be modified to take into account the subdirectory, or Klein itself patched to do so.
3. Links in _app/views_ must be modified to take into account the subdirectory.

### PR DESCRIPTION
**Pull request by TakeMeNL**
The initial issue was to run PufferPanel in a subdirectory, This is very challenging with Klein.
But did manage to fix the issue.

The "standalone", non sub-directory, version will still work like normal.

If you want to run PufferPanel in a sub-directory you need to put the contents of public in the directory you want. Example `http://domain.tld/pufferpanel/`.
The rest of the software needs to live OUTSIDE the public directory.
If you want multi sub-directories this is possible to `http://domain.tld/test/test/`

```
domain.tld
|-- public_html
|   `-- pufferpanel
|       |-- .htaccess
|       |-- index.php
|       `-- router.php
`-- pufferpanel
    `-- pufferpanel git project.
```

This fixes #687

### CODE REVIEW

**Comment by TakeMeNL** [comment]
Fixed MR Conflict

File: public/index.php [hunk 1]
```diff
@@ -1,2 +1,3 @@
 <?php
-header('Location: /index');
+$baseUrl = !file_exists(dirname(__FILE__).'/pufferpanel') ? dirname($_SERVER['PHP_SELF']).'/' : '/';
```
Comment (TakeMeNL) [inline public/index.php]: I assumed you wanted this file. If it's up to me it's gone.

File: src/core/ppresponse.php [hunk 2]
```diff
@@ -41,7 +41,10 @@ public function code($code = null)
     public function redirect($url, $code = 302)
     {
         $this->setAltered();
-        parent::redirect($url, $code);
+        if (strpos($url,'http') !== false)
```
Comment (TakeMeNL) [inline src/core/ppresponse.php]: Can do

File: public/router.php [hunk 3]
```diff
@@ -25,15 +25,21 @@
 	date_default_timezone_set('UTC');
 }
 
-define('BASE_DIR', dirname(__DIR__).'/');
+$baseDir = str_replace('\\', '/', dirname(dirname(__FILE__)).'/');
+if (!is_file($baseDir.'pufferpanel')) {
+    $baseUrl = dirname($_SERVER['PHP_SELF']);
```
Comment (TakeMeNL) [inline public/router.php]: Yeah sure but the master url must always exists then and filled in. It is a possibility to.  
BUT you still need to know where your files are. If you can tell me how you want to do that. I build it. For now this is the best and easiest solution.

Comment (LordRalex) [inline public/router.php]: Master url will always exist and be filled in.
This is not the best nor easiest solution. 

We will deal with making master url required. All that really matters there is the subfolder, which you can assume it's / if there's no master url.

Comment (LordRalex) [inline public/router.php]: And why is knowing how the files are structured required? I need a technical reason for this.

Comment (TakeMeNL) [inline public/router.php]: Personally was the problem you don't want you source code in root of you're website. But you could fix this with nginx and apache. But that not a easy install. This way its just drop and go.

Comment (LordRalex) [inline public/router.php]: We already ship configs that do just that. This solution is not drop and go if you have to change folder structures to get it to work. I do not get why that is needed.

I *prefer* just changing configs. If it fails, you restore a config and it works. Moving folders around *WILL BREAK* when we do updates. This solution does not fully solve the issue. 

We need a way to solve this using just the nginx/apache configs and the master_url. Using any form of "folder layout" will be a nightmare for supporting setups.

Comment (TakeMeNL) [inline public/router.php]: Oke well then we have our answer, will close th PR

**Comment by LordRalex** [comment]
<Bridge> T​akeMeNL | If you download puffer panel and put it in my existing website `domain.tld/members/panel`. What do you put there?
<Bridge> T​akeMeNL | If i put all of pufferpanel in `domain.tld/members/panel` the web root to acces puffer would be `domain.tld/members/panel/web` And thats stupid. with .htaccess you're not resolving that issue
<Bridge> T​akeMeNL | And correct me if am wrong but i dont see how to fix that.

<Lord_Ralex> How is it going to be /web, why is the .htaccess not resolving that.

Even though we can't re-open this issue, you're not explaining *why* that's the case. It does not make any sense to me as a developer why this setup requires some specialized folder structure to work. No program should be doing stuff like that unless it's very clear. Meteor does and it's actually broken if you do it right.

Apache supports configurations and expects them to handle this stuff, so how is the .htaccess NOT handling that.

If I say domain.tld/members/panel is my panel, the .htaccess WILL redirect ALL calls to that path to the router.php, which handles the magic to turn it into our calls. Why is some crazy folder structure then needed to support that? Unless I cannot tell, Apache should be able to support that, they seem to point to Alias for stuff.

This also is what phpadmin does, it uses an alias.
Alias /phpmyadmin /usr/share/phpmyadmin

**Pull request closed without merging**